File: Depth_First_Search/1559.Detect_Cycles_in_2D_Grid/C++/1559.Detect_Cycles_in_2D_Grid.cpp
class Solution {
public:
    bool containsCycle(vector<vector<char>> & grid) 
    {
        //=============================================//
        // Input type:                                 //
        //   - vector<vector<char>> ref. (grid vector) //
        // Return type:                                //
        //   - bool (result flag)                      //
        //=============================================//

        /*Initialize*/
        ///// Length of rows, columns /////
        int len_rows = grid.size(), len_cols = (grid[0]).size();

        ///// Record dictionary, hashmap /////
        unordered_map<int, unordered_map<int, bool>> record_map;


        /*Recursion-based DFS loop traversal*/
        for (int rows_idx = 0; (rows_idx < len_rows); (rows_idx++)) //Rows
        {
            for (int cols_idx = 0; (cols_idx < len_cols); (cols_idx++)) //Columns
            {
                ///// Check if the current indexed-char matched consitions or not /////
                if ((grid[rows_idx])[cols_idx] != '0')
                {
                    ///// Check if the current flag matched conditions or not /////
                    if (dfsTraversal(grid, len_rows, len_cols, rows_idx, cols_idx, (grid[rows_idx])[cols_idx], (-1), record_map))
                    {
                        return true;
                    }
                    else { ; }

                    record_map.clear(); //Update, Reset
                }
                else { ; }

            } //Columns

        } //Rows

        return false;
    }


    bool dfsTraversal(vector<vector<char>> & grid, int & len_rows, int & len_cols, int rows_idx, int cols_idx, char record_char, int record_dirc, unordered_map<int, unordered_map<int, bool>> & record_map)
    {
        //====================================================================================//
        // Input type:                                                                        //
        //   - vector<vector<char>> ref. (grid vector)                                        //
        //   - int ref. (length of rows)                                                      //
        //   - int ref. (length of columns)                                                   //
        //   - int (rows index)                                                               //
        //   - int (columns index)                                                            //
        //   - char (record char)                                                             //
        //   - int (record durection)                                                         //
        //   - unordered_map<int, unordered_map<int, bool>> ref. (record dictionary, hashmap) //
        // Return type:                                                                       //
        //   - bool (result flag)                                                             //
        //====================================================================================//
        if (((rows_idx < 0) || (rows_idx >= len_rows)) || ((cols_idx < 0) || (cols_idx >= len_cols))) { return false; } //Issue, Boundary-case handling

        if (record_map.find(rows_idx) != record_map.end()) //Issue, Boundary-case handling
        {
            if ((record_map[rows_idx]).find(cols_idx) != (record_map[rows_idx]).end()) { return true; }
        }

        if ((grid[rows_idx])[cols_idx] != record_char) { return false; } //Issue, Boundary-case handling


        /*Whole process, flow*/
        (record_map[rows_idx])[cols_idx] = true; //Update, Record

        (grid[rows_idx])[cols_idx] = '0'; //Update, Overwrite

        ///// Check if the current directions matched conditions or not /////
        if (record_dirc == (-1)) //Whole
        {
            return (dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx - 1), record_char, 0, record_map) | dfsTraversal(grid, len_rows, len_cols, (rows_idx - 1), cols_idx, record_char, 1, record_map) | dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx + 1), record_char, 2, record_map) | dfsTraversal(grid, len_rows, len_cols, (rows_idx + 1), cols_idx, record_char, 3, record_map)); //Recursion function call
        }

        else if (record_dirc == 0) //Left
        {
            return (dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx - 1), record_char, 0, record_map) | dfsTraversal(grid, len_rows, len_cols, (rows_idx - 1), cols_idx, record_char, 1, record_map) | dfsTraversal(grid, len_rows, len_cols, (rows_idx + 1), cols_idx, record_char, 3, record_map)); //Recursion function call
        }

        else if (record_dirc == 1) //Up
        {
            return (dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx - 1), record_char, 0, record_map) | dfsTraversal(grid, len_rows, len_cols, (rows_idx - 1), cols_idx, record_char, 1, record_map) | dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx + 1), record_char, 2, record_map)); //Recursion function call
        }

        else if (record_dirc == 2) //Right
        {
            return (dfsTraversal(grid, len_rows, len_cols, (rows_idx - 1), cols_idx, record_char, 1, record_map) | dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx + 1), record_char, 2, record_map) | dfsTraversal(grid, len_rows, len_cols, (rows_idx + 1), cols_idx, record_char, 3, record_map)); //Recursion function call
        }

        return (dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx - 1), record_char, 0, record_map) | dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx + 1), record_char, 2, record_map) | dfsTraversal(grid, len_rows, len_cols, (rows_idx + 1), cols_idx, record_char, 3, record_map)); //Recursion function call
    }
};

File: Depth_First_Search/1559.Detect_Cycles_in_2D_Grid/Python/1559.Detect_Cycles_in_2D_Grid.py
class Solution(object):
    def containsCycle(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: bool
        """
        #======================================#
        # Recursion-based DFS traversal method #
        #======================================#

        ############
        #Initialize
        ##### Length of rows, columns #####
        len_rows, len_cols = len(grid), len(grid[0])

        ##### Record dictionary/hashmap #####
        record_dict = {}


        ####################################
        #Recursion-based DFS loop traversal
        for rows_idx in range(len_rows):
            for cols_idx in range(len_cols):

                ##### Check if the current indexed-char matched conditions or not #####
                if ((grid[rows_idx])[cols_idx] != "0"):
                    if (self.dfsTraversal(grid, len_rows, len_cols, rows_idx, cols_idx, (grid[rows_idx])[cols_idx], (-1), record_dict)): #Recursion function call
                        return True
                    else:
                        pass

                    record_dict.clear() #Update/Reset

                else:
                    pass

        return False


    def dfsTraversal(self, grid, len_rows, len_cols, rows_idx, cols_idx, record_char, record_dirc, record_dict):
        """
        :type grid: List[List[str]]
        :type len_rows: int
        :type len_cols: int
        :type rows_idx: int
        :type cols_idx: int
        :type record_char: char
        :type record_dirc: int
        :type record_dict: dict
        :rtype: bool
        """
        #========================================#
        # Recursion-based DFS traversal function #
        #========================================#
        if (((rows_idx < 0) or (rows_idx >= len_rows)) or ((cols_idx < 0) or (cols_idx >= len_cols))): #Issue/Boundary-case handling
            return False

        if ((rows_idx, cols_idx) in record_dict): #Issue/Boundary-case handling
            return True

        if ((grid[rows_idx])[cols_idx] != record_char): #Issue/Boundary-case handling
            return False


        ###################
        #Whole process/flow
        record_dict[(rows_idx, cols_idx)] = True #Update/Record

        (grid[rows_idx])[cols_idx] = "0" #Update/Overwrite

        ##### Check which previous direction is #####
        if (record_dirc == (-1)): #Original
            return (self.dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx - 1), record_char, 0, record_dict) | self.dfsTraversal(grid, len_rows, len_cols, (rows_idx - 1), cols_idx, record_char, 1, record_dict) | self.dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx + 1), record_char, 2, record_dict) | self.dfsTraversal(grid, len_rows, len_cols, (rows_idx + 1), cols_idx, record_char, 3, record_dict)) #Recursion function call

        elif (record_dirc == 0): #Left
            return (self.dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx - 1), record_char, 0, record_dict) | self.dfsTraversal(grid, len_rows, len_cols, (rows_idx - 1), cols_idx, record_char, 1, record_dict) | self.dfsTraversal(grid, len_rows, len_cols, (rows_idx + 1), cols_idx, record_char, 3, record_dict)) #Recursion function call

        elif (record_dirc == 1): #Up
            return (self.dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx - 1), record_char, 0, record_dict) | self.dfsTraversal(grid, len_rows, len_cols, (rows_idx - 1), cols_idx, record_char, 1, record_dict) | self.dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx + 1), record_char, 2, record_dict)) #Recursion function call

        elif (record_dirc == 2): #Right
            return (self.dfsTraversal(grid, len_rows, len_cols, (rows_idx - 1), cols_idx, record_char, 1, record_dict) | self.dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx + 1), record_char, 2, record_dict) | self.dfsTraversal(grid, len_rows, len_cols, (rows_idx + 1), cols_idx, record_char, 3, record_dict)) #Recursion function call

        elif (record_dirc == 3): #Down
            return (self.dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx - 1), record_char, 0, record_dict) | self.dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx + 1), record_char, 2, record_dict) | self.dfsTraversal(grid, len_rows, len_cols, (rows_idx + 1), cols_idx, record_char, 3, record_dict)) #Recursion function call
